packages/trie/src/verifyRangeProof.ts

```diff
@@ -0,0 +1,500 @@
+ *
+ * - Zero element proof. In this case a single non-existent proof is enough to prove.
+ *   Besides, if there are still some other leaves available on the right side, then
+ *   an error will be returned.
+ *
+ * - Two edge elements proof. In this case two existent or non-existent proof(fisrt and last) should be provided.
+ *
+ * NOTE: Currently only supports verification when the length of firstKey and lastKey are the same.
+ *
+ * @param rootHash - root hash.
+ * @param firstKey - first key.
+ * @param lastKey - last key.
+ * @param keys - key list.
+ * @param values - value list, one-to-one correspondence with keys.
+ * @param proof - proof node list.
+ * @returns a flag to indicate whether there exists more trie node in the trie
+ */
+export async function verifyRangeProof(
+  rootHash: Buffer,
+  firstKey: Nibbles | null,
+  lastKey: Nibbles | null,
+  keys: Nibbles[],
+  values: Buffer[],
+  proof: Buffer[] | null
+) {
+  if (keys.length !== values.length) {
+    throw new Error('invalid keys and values')
+  }
+
+  // Make sure the keys are in order
+  for (let i = 0; i < keys.length - 1; i++) {
+    if (nibblesCompare(keys[i], keys[i + 1]) >= 0) {
+      throw new Error('invalid keys')
+    }
+  }
+  // Make sure all values are present
+  for (const value of values) {
+    if (value.length === 0) {
+      throw new Error('invalid values')
+    }
+  }
+
+  // All elements proof
+  if (proof === null) {
+    const trie = new Trie()
+    for (let i = 0; i < keys.length; i++) {
+      await trie.put(nibblesToBuffer(keys[i]), values[i])
+    }
+    if (rootHash.compare(trie.root) !== 0) {
+      throw new Error('invalid proof')
+    }
+    return false
+  }
+
+  if (firstKey === null || lastKey === null) {
+    throw new Error('invalid first or last key')
+  }
+
+  // Zero element proof
+  if (keys.length === 0) {
+    const { trie, value } = await verifyProof(rootHash, nibblesToBuffer(firstKey), proof)
+
+    if (value !== null || (await hasRightElement(trie, firstKey))) {
+      throw new Error('invalid proof')
+    }
+
+    return false
+  }
+
+  // One element proof
+  if (keys.length === 1 && nibblesCompare(firstKey, lastKey) === 0) {
+    const { trie, value } = await verifyProof(rootHash, nibblesToBuffer(firstKey), proof)
+
+    if (nibblesCompare(firstKey, keys[0]) !== 0) {
+      throw new Error('invalid proof')
+    }
+    if (value === null || value.compare(values[0]) !== 0) {
+      throw new Error('invalid proof')
+    }
+
+    return hasRightElement(trie, firstKey)
+  }
+
+  // Two edge elements proof
+
+  if (nibblesCompare(firstKey, lastKey) >= 0) {
+    throw new Error('invalid proof')
+  }
+  if (firstKey.length !== lastKey.length) {
+    throw new Error('invalid proof')
+  }
+
+  let trie = new Trie(null, rootHash)
+  trie = await Trie.fromProof(proof, trie)
+
+  // Remove all nodes between two edge proofs
+  const empty = await unsetInternal(trie, firstKey, lastKey)
+
```

Comment:
I think he's suggesting to trim the empty newline, I think I would agree here it would be fine to group the empty logic.

Reply:
> I think he's suggesting to trim the extra newline, I think I would agree here it would be fine to group the empty logic.

Yes, trim in `b7326ac2c1ba535b4b84fc2d50c7c313ebedf103`